File: HomelabSwift/Homelab/Stores/TenantStore.swift
import Foundation
import Observation

/// Persists the device-local `TenantSelection` (configured tenants + active selection) as a single
/// JSON blob in `UserDefaults`. All mutation rules live in `TenantSelection`; this type only stores
/// the result and generates ids for new tenants.
///
/// A device that never adds a second tenant keeps `TenantSelection.initial` and never writes the
/// key, so a single-tenant install is byte-identical to pre-Phase-4.
@Observable
@MainActor
final class TenantStore {

    private(set) var selection: TenantSelection

    @ObservationIgnored private let defaults: UserDefaults
    @ObservationIgnored private static let storageKey = "tenant_selection_v1"

    init(defaults: UserDefaults = .standard) {
        self.defaults = defaults
        self.selection = Self.load(from: defaults)
    }

    var activeTenant: Tenant { selection.activeTenant }
    var activeTenantRef: String { selection.activeTenantId }
    var isSingleTenant: Bool { selection.isSingleTenant }
    var membershipRefs: Set<String> { selection.membershipRefs }

    /// Creates a tenant with a fresh id and returns the stored selection.
    @discardableResult
    func addTenant(name: String, kind: TenantKind = .customer) -> TenantSelection {
        let tenant = Tenant(
            id: "tenant-\(UUID().uuidString.lowercased())",
            name: name.trimmingCharacters(in: .whitespacesAndNewlines),
            kind: kind
        )
        return apply { $0.adding(tenant) }
    }

    @discardableResult
    func renameTenant(id: String, name: String) -> TenantSelection {
        apply { $0.renaming(id: id, to: name.trimmingCharacters(in: .whitespacesAndNewlines)) }
    }

    @discardableResult
    func removeTenant(id: String) -> TenantSelection {
        apply { $0.removing(id: id) }
    }

    @discardableResult
    func setActiveTenant(id: String) -> TenantSelection {
        apply { $0.activating(id: id) }
    }

    @discardableResult
    func setAllTenantsMode(_ enabled: Bool) -> TenantSelection {
        apply { $0.settingAllTenantsMode(enabled) }
    }

    private func apply(_ transform: (TenantSelection) -> TenantSelection) -> TenantSelection {
        let next = transform(selection).normalized()
        selection = next
        persist(next)
        return next
    }

    private func persist(_ selection: TenantSelection) {
        guard let data = try? JSONEncoder().encode(selection) else { return }
        defaults.set(data, forKey: Self.storageKey)
    }

    private static func load(from defaults: UserDefaults) -> TenantSelection {
        guard
            let data = defaults.data(forKey: storageKey),
            let decoded = try? JSONDecoder().decode(TenantSelection.self, from: data)
        else {
            return .initial
        }
        return decoded.normalized()
    }
}

File: HomelabAndroid/app/src/main/java/com/homelab/app/data/local/TenantStore.kt
package com.homelab.app.data.local

import androidx.datastore.core.DataStore
import androidx.datastore.preferences.core.Preferences
import androidx.datastore.preferences.core.edit
import androidx.datastore.preferences.core.stringPreferencesKey
import com.homelab.app.domain.model.Tenant
import com.homelab.app.domain.model.TenantKind
import com.homelab.app.domain.model.TenantSelection
import java.util.UUID
import javax.inject.Inject
import javax.inject.Singleton
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.flow.map
import kotlinx.serialization.json.Json

/**
 * Persists the device-local [TenantSelection] (configured tenants + active selection) in the
 * Preferences DataStore under a single JSON key. All mutation rules live in [TenantSelection];
 * this class only stores the result and generates ids for new tenants.
 *
 * A device that never adds a second tenant keeps the default [TenantSelection.INITIAL] and never
 * writes the key, so a single-tenant install is byte-identical to pre-Phase-4.
 */
@Singleton
class TenantStore @Inject constructor(
    private val dataStore: DataStore<Preferences>,
    private val json: Json
) {
    private val key = stringPreferencesKey(STORE_KEY)

    val selection: Flow<TenantSelection> = dataStore.data.map { decode(it[key]) }

    suspend fun current(): TenantSelection = decode(dataStore.data.first()[key])

    /** Creates a tenant with a fresh id and returns the stored selection. */
    suspend fun addTenant(name: String, kind: TenantKind = TenantKind.CUSTOMER): TenantSelection {
        val tenant = Tenant(id = "tenant-${UUID.randomUUID()}", name = name.trim(), kind = kind)
        return update { it.adding(tenant) }
    }

    suspend fun renameTenant(id: String, name: String): TenantSelection =
        update { it.renaming(id, name.trim()) }

    suspend fun removeTenant(id: String): TenantSelection = update { it.removing(id) }

    suspend fun setActiveTenant(id: String): TenantSelection = update { it.activating(id) }

    suspend fun setAllTenantsMode(enabled: Boolean): TenantSelection =
        update { it.settingAllTenantsMode(enabled) }

    private suspend fun update(transform: (TenantSelection) -> TenantSelection): TenantSelection {
        var result = TenantSelection.INITIAL
        dataStore.edit { preferences ->
            result = transform(decode(preferences[key])).normalized()
            preferences[key] = json.encodeToString(TenantSelection.serializer(), result)
        }
        return result
    }

    private fun decode(raw: String?): TenantSelection {
        if (raw.isNullOrBlank()) return TenantSelection.INITIAL
        return runCatching {
            json.decodeFromString(TenantSelection.serializer(), raw).normalized()
        }.getOrDefault(TenantSelection.INITIAL)
    }

    private companion object {
        const val STORE_KEY = "tenant_selection_v1"
    }
}
